### ISSUE TITLE
Bug: This gem messes with ExecJS runtime selection

### ISSUE THREAD

**Issue by glebm** [ccorn/sskatex#2]
This causes issues when trying to use it alongside anything else that relies on ExecJS automatic runtime selection. This is plenty of gems and user code, including every default Rails installation (e.g. via uglifier).

https://github.com/ccorn/sskatex/blame/96a7320284acfa8ca0f55f36e6cb861f482761b7/lib/sskatex.rb#L63-L69

I'm trying to add a `katex` engine to kramdown (https://github.com/gettalong/kramdown/pull/489) but I get test errors such as:

```
TestFiles#test_/home/glebm/repos/glebm/kramdown/test/testcases/span/math/katex_text_to_html:
ExecJS::Error: ExecJS disabled
    /home/glebm/.rvm/gems/ruby-2.5.0/gems/execjs-2.7.0/lib/execjs/disabled_runtime.rb:18:in `compile'
    /home/glebm/.rvm/gems/ruby-2.5.0/gems/execjs-2.7.0/lib/execjs/module.rb:27:in `compile'
    /home/glebm/.rvm/gems/ruby-2.5.0/gems/katex-0.4.0.beta/lib/katex.rb:47:in `block in katex_context'
    /home/glebm/.rvm/gems/ruby-2.5.0/gems/katex-0.4.0.beta/lib/katex.rb:46:in `synchronize'
    /home/glebm/.rvm/gems/ruby-2.5.0/gems/katex-0.4.0.beta/lib/katex.rb:46:in `katex_context'
    /home/glebm/.rvm/gems/ruby-2.5.0/gems/katex-0.4.0.beta/lib/katex.rb:30:in `render'
    /home/glebm/repos/glebm/kramdown/lib/kramdown/converter/math_engine/katex.rb:20:in `call'
    /home/glebm/repos/glebm/kramdown/lib/kramdown/converter/base.rb:221:in `format_math'
```

**Comment by ccorn** [comment]
Your gem uses `ExecJS.compile` which tries to setup and use a global ExecJS runtime. This forces one ExecJS user's runtime choice upon all other users, which is obviously bad.

ExecJS supports an object model. Create an `ExecJS::Runtime` object, e.g. `ExecJS::Runtimes.best_available` and let that one `.compile` stuff.

**Comment by glebm** [comment]
Thanks for the tip, I'll fix my gem.

Why is your gem changing `ExecJS` default behaviour? Do you agree that it shouldn't?

**Comment by ccorn** [comment]
Because autodetection should only take place when users do not request something else. SsKaTeX supports user-specified engine choices (so we do not need to bother about cases where installation of one new shiny engine suddenly breaks KaTeX). The way ExecJS works is by first looking into the environment variable `EXECJS_RUNTIME`, trying to find a same-named `Runtime` object, then using the `.best_available` method which involves filesystem tests (and perhaps more, but the more expensive variants have been deprecated). Before first use it then tries to compile an empty script if I recall correctly. All this adds overhead and potential vulnerabilites. And this takes place immediately when loading ExecJS, so to defer that sort of behavior until user options have been checked, I temporarily set `EXECJS_RUNTIME=Disabled`. (But I restore that variable after loading `ExecJS` and emulate the ExecJS behavior if nothing else has been requested.)

**Comment by glebm** [comment]
Right, however, restoring it doesn't work for the global `ExecJS.runtime`, because it is initialized immediately on `require 'execjs'`:

https://github.com/rails/execjs/blob/5f78865d36976f85b42e8529dec272285d8f3b85/lib/execjs.rb#L4-L6

If the user relies on the global runtime + autodetection (the easiest way to use ExecJS), simply requireing `sskatex` breaks their code.

Since you're not using the global runtime, why do you care if autodetection is performed?

If you care about a particular application, consider that it can be disabled there rather than modifying the default behaviour of a very common dependency.

**Comment by glebm** [comment]
I think I've found the perfect solution for the `sskatex` gem.

Instead of this:

```ruby
require 'execjs'
```

Do this:

```ruby
require 'execjs/module'
require 'execjs/runtimes'
```

This way you don't need any hacks, don't break other gems, and avoid needless autodetection.

**Comment by ccorn** [comment]
Relying on globals adds interdependencies that break other stuff.
Nevertheless, not using SsKaTeX should not result in breaks.
If normal kramdown operations work with your math engine, then the issue is that in kramdown's `test/test_files.rb` we explicitly test for SsKaTeX availabilty by, among other things, requiring it. That sets the global runtime to disabled.

If you really want to use the global ExecJS even in kramdown where you cannot control all sources, then you could use `ExecJS.runtime = ExecJS::Runtimes.autodetect` explicitly after loading. I'd consider that a kludge however. The best approach is to avoid globals.

**Comment by glebm** [comment]
> Relying on globals adds interdependencies that break other stuff.

I respectfully disagree. Reading from globals is fine. It's writing to shared globals from gems that breaks stuff. In particular, setting global environment variables from a gem.

I'm glad we've found a fix that avoids this (PR sent).

**Comment by ccorn** [comment]
I have merged #3 and prepare a new release. Thanks for the pull request.
